packages/muelu/src/Utils/MueLu_AlgebraicPermutationStrategy_def.hpp

```diff
@@ -31,674 +31,690 @@
-      procMultRequestedColIds[nMyOffset + i] = multipleColRequests[i]; // all weights are > 0 ?
+    // communicate how often each column index is requested by the different procs
+    gDomVec->doExport(*gColVec,*exporter,Xpetra::ADD);
+    gColVec->doImport(*gDomVec,*exporter,Xpetra::INSERT); // probably not needed // TODO check me
+
+    //*****************************************************************************************
+    // first communicate ALL global ids of column indices which are requested by more
+    // than one proc to all other procs
+    // detect which global col indices are requested by more than one proc
+    // and store them in the multipleColRequests vector
+    std::vector<GlobalOrdinal> multipleColRequests; // store all global column indices from current processor that are also
+    // requested by another processor. This is possible, since they are stored
+    // in gDomVec which is based on the nonoverlapping domain map. That is, each
+    // global col id is handled by exactly one proc.
+    std::queue<GlobalOrdinal> unusedColIdx; // unused column indices on current processor
+
+    for(size_t sz = 0; sz < gDomVec->getLocalLength(); ++sz) {
+      Teuchos::ArrayRCP< const Scalar > arrDomVec = gDomVec->getData(0);
+      //
+      // FIXME (mfh 30 Oct 2015) I _think_ it's OK to check just the
+      // real part, because this is a count.  (Shouldn't MueLu use
+      // mag_type instead of Scalar here to save space?)
+      //
+      if(Teuchos::ScalarTraits<Scalar>::real(arrDomVec[sz]) > MT_ONE) {
+        multipleColRequests.push_back(gDomVec->getMap()->getGlobalElement(sz));
+      } else if(Teuchos::ScalarTraits<Scalar>::real (arrDomVec[sz]) == MT_ZERO) {
+        unusedColIdx.push(gDomVec->getMap()->getGlobalElement(sz));
+      }
     }
 
-    // template ordinal, package (double)
-    Teuchos::reduceAll(*comm, Teuchos::REDUCE_MAX, Teuchos::as<int>(globalMultColRequests), &procMultRequestedColIds[0], &global_procMultRequestedColIds[0]);
+    // communicate the global number of column indices which are requested by more than one proc
+    LocalOrdinal localMultColRequests  = Teuchos::as<LocalOrdinal>(multipleColRequests.size());
+    LocalOrdinal globalMultColRequests = 0;
+
+    // sum up all entries in multipleColRequests over all processors
+      //
+      // FIXME (lbv 11 Feb 2019) why cast localMultColRequests as LocalOrdinal
+      // when it was properly declared as such just a line above?
+      //
+    MueLu_sumAll(gDomVec->getMap()->getComm(), (LocalOrdinal)localMultColRequests, globalMultColRequests);
+
+    if(globalMultColRequests > 0) {
+      // special handling: two processors request the same global column id.
+      // decide which processor gets it
+
+      // distribute number of multipleColRequests to all processors
+      // each processor stores how many column ids for exchange are handled by the cur proc
+      std::vector<GlobalOrdinal> numMyMultColRequests(numProcs,0);
```

Comment:
There must be a better way to communicate this information than an all-reduce on a whole array, but that's a separate issue.  (Maybe Distributor already encapsulates this communication pattern for you?)